cmd/kubeadm/app/cmd/config.go

```diff
@@ -0,0 +1,173 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/renstrom/dedent"
+	"github.com/spf13/cobra"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmapiext "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/uploadconfig"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+func NewCmdConfig(out io.Writer) *cobra.Command {
+
+	var kubeConfigFile string
+	cmd := &cobra.Command{
+		Use:   "config",
+		Short: "Manage configuration for a kubeadm cluster persisted in a ConfigMap in the cluster.",
+		Long: dedent.Dedent(`
+			There is a ConfigMap in the %s namespace called %q that kubeadm uses to store internal configuration about the
+			cluster. kubeadm CLI v1.8.0+ automatically creates this ConfigMap with used config on 'kubeadm init', but if you
+			initialized your cluster using kubeadm v1.7.x or lower, you must use the 'config upload' command to create this
+			ConfigMap in order for 'kubeadm upgrade' to be able to configure your upgraded cluster correctly.
+		`),
+		// Without this callback, if a user runs just the "upload"
+		// command without a subcommand, or with an invalid subcommand,
+		// cobra will print usage information, but still exit cleanly.
+		// We want to return an error code in these cases so that the
+		// user knows that their command was invalid.
+		RunE: subCmdRunE("config"),
+	}
+
+	cmd.PersistentFlags().StringVar(&kubeConfigFile, "kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use for talking to the cluster")
+
+	cmd.AddCommand(NewCmdConfigUpload(out, &kubeConfigFile))
+	cmd.AddCommand(NewCmdConfigView(out, &kubeConfigFile))
+
+	return cmd
+}
+
+func NewCmdConfigUpload(out io.Writer, kubeConfigFile *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "upload",
+		Short: "Upload configuration about the current state so 'kubeadm upgrade' later can know how to configure the upgraded cluster",
+		RunE:  subCmdRunE("upload"),
+	}
+
+	cmd.AddCommand(NewCmdConfigUploadFromFile(out, kubeConfigFile))
+	cmd.AddCommand(NewCmdConfigUploadFromFlags(out, kubeConfigFile))
+	return cmd
+}
+
+func NewCmdConfigView(out io.Writer, kubeConfigFile *string) *cobra.Command {
+	return &cobra.Command{
+		Use:   "view",
+		Short: "View the kubeadm configuration stored inside the cluster",
```

Comment:
Please denote details of where the configuration is stored so they can independently verify.